File: app/src/main/java/social/entourage/android/user/edit/UserEditFragment.java
package social.entourage.android.user.edit;

import android.content.ActivityNotFoundException;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import androidx.annotation.NonNull;
import androidx.core.app.NotificationManagerCompat;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import com.squareup.otto.Subscribe;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import social.entourage.android.MainActivity;
import social.entourage.android.EntourageActivity;
import social.entourage.android.EntourageApplication;
import social.entourage.android.EntourageComponent;
import social.entourage.android.EntourageEvents;
import social.entourage.android.R;
import social.entourage.android.api.model.BaseOrganization;
import social.entourage.android.api.model.User;
import social.entourage.android.api.tape.Events;
import social.entourage.android.authentication.AuthenticationController;
import social.entourage.android.authentication.UserPreferences;
import social.entourage.android.base.EntourageDialogFragment;
import social.entourage.android.partner.PartnerFragment;
import social.entourage.android.tools.BusProvider;
import social.entourage.android.tools.CropCircleTransformation;
import social.entourage.android.user.UserFragment;
import social.entourage.android.user.UserOrganizationsAdapter;
import social.entourage.android.user.edit.partner.UserEditPartnerFragment;
import social.entourage.android.user.edit.photo.PhotoChooseSourceFragment;
import timber.log.Timber;

public class UserEditFragment extends EntourageDialogFragment implements UserEditActionZoneFragment.FragmentListener {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = "user_edit_fragment";

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @Inject
    UserEditPresenter presenter;

    @BindView(R.id.scrollView)
    ScrollView scrollView;

    private int scrollViewY = 0;

    @BindView(R.id.user_photo)
    ImageView userPhoto;

    @BindView(R.id.user_edit_firstname)
    TextView userFirstname;

    @BindView(R.id.user_edit_lastname)
    TextView userLastname;

    @BindView(R.id.user_email)
    TextView userEmail;

    @BindView(R.id.user_phone)
    TextView userPhone;

    @BindView(R.id.user_about)
    TextView userAbout;

    @BindView(R.id.user_notifications_image)
    ImageView userNotificationsStatusImage;

    @BindView(R.id.user_associations_view)
    RecyclerView userAssociationsView;

    @BindView(R.id.user_associations_layout)
    View userAssociationLayout;

    @BindView(R.id.user_action_zone)
    TextView userActionZone;

    @BindView(R.id.user_action_zone_button)
    Button userActionZoneButton;

    @BindView(R.id.user_edit_progressBar)
    ProgressBar progressBar;

    private UserOrganizationsAdapter organizationsAdapter;

    private User editedUser;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public UserEditFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View v = inflater.inflate(R.layout.fragment_user_edit, container, false);
        ButterKnife.bind(this, v);
        EntourageEvents.logEvent(EntourageEvents.EVENT_SCREEN_09_2);
        return v;
    }

    @Override
    public void onViewCreated(@NonNull final View view, final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupComponent(EntourageApplication.get(getActivity()).getEntourageComponent());
        configureView();
    }

    protected void setupComponent(EntourageComponent entourageComponent) {
        DaggerUserEditComponent.builder()
                .entourageComponent(entourageComponent)
                .userEditModule(new UserEditModule(this))
                .build()
                .inject(this);
    }

    @Override
    public void onStart() {
        super.onStart();

        configureNotifications();

        BusProvider.getInstance().register(this);
    }

    @Override
    public void onStop() {
        super.onStop();

        BusProvider.getInstance().unregister(this);
    }

    public void configureView() {
        if (getActivity() != null) {
            if (editedUser == null) {
                User me = EntourageApplication.me(getActivity());
                if (me == null) return;
                editedUser = me.clone();
                if (editedUser == null) return;
            }

            if (editedUser.getAvatarURL() != null) {
                Picasso.get().load(Uri.parse(editedUser.getAvatarURL()))
                        .placeholder(R.drawable.ic_user_photo)
                        .transform(new CropCircleTransformation())
                        .into(userPhoto);
            }
            else {
                Picasso.get().load(R.drawable.ic_user_photo)
                        .transform(new CropCircleTransformation())
                        .into(userPhoto);
            }

            userFirstname.setText(editedUser.getFirstName());
            userLastname.setText(editedUser.getLastName());
            userEmail.setText(editedUser.getEmail());
            userPhone.setText(editedUser.getPhone());
            userAbout.setText(editedUser.getAbout());

            List<BaseOrganization> organizationList = new ArrayList<>();
            if (editedUser.getPartner() != null) {
                organizationList.add(editedUser.getPartner());
            }
            if (editedUser.getOrganization() != null) {
                organizationList.add(editedUser.getOrganization());
            }

            if(organizationList.size() > 0) {
                if (organizationsAdapter == null) {
                    userAssociationsView.setLayoutManager(new LinearLayoutManager(getActivity()));
                    organizationsAdapter = new UserOrganizationsAdapter(organizationList);
                    userAssociationsView.setAdapter(organizationsAdapter);
                } else {
                    organizationsAdapter.setOrganizationList(organizationList);
                }
                userAssociationLayout.setVisibility(View.VISIBLE);
            }

            User.Address address = editedUser.getAddress();
            if (address == null) {
                userActionZone.setText("");
                userActionZoneButton.setText(R.string.user_edit_action_zone_button_no_address);
            } else {
                String displayAddress = address.getDisplayAddress();
                if (displayAddress == null || displayAddress.length() == 0) {
                    userActionZone.setText("");
                } else {
                    userActionZone.setText(getString(R.string.user_edit_action_zone_format, address.getDisplayAddress()));
                }
                userActionZoneButton.setText(R.string.user_edit_action_zone_button);
            }
        }
    }

    public User getEditedUser() {
        return editedUser;
    }

    public void displayToast(String message) {
        if (getActivity() != null) {
            Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
        }
    }

    public void saveNewPassword(String newPassword) {
        if (presenter != null) {
            progressBar.setVisibility(View.VISIBLE);
            presenter.saveNewPassword(newPassword);
        }
    }

    // ----------------------------------
    // Buttons Handling
    // ----------------------------------

    @OnClick(R.id.title_close_button)
    protected void onCloseButtonClicked() {
        dismiss();
    }

    @OnClick({R.id.user_firstname_layout, R.id.user_lastname_layout})
    protected void onEditFirstname() {
        showEditProfile(UserEditProfileFragment.EDIT_NAME);
    }

    @OnClick(R.id.user_email_layout)
    protected void onEditEmail() {
        showEditProfile(UserEditProfileFragment.EDIT_EMAIL);
    }

    @OnClick(R.id.user_password_layout)
    protected void onEditPassword() {
        UserEditPasswordFragment fragment = new UserEditPasswordFragment();
        fragment.show(getFragmentManager(), UserEditPasswordFragment.TAG);
    }

    @OnClick(R.id.user_about_edit_button)
    protected void onEditAboutClicked() {
        scrollViewY = scrollView.getScrollY();
        UserEditAboutFragment editAboutFragment = new UserEditAboutFragment();
        editAboutFragment.show(getFragmentManager(), UserEditAboutFragment.TAG);
    }

    @OnClick(R.id.user_delete_account_button)
    protected void onDeleteAccountClicked() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setMessage(R.string.user_delete_account_dialog)
                .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(final DialogInterface dialog, final int which) {
                        if (presenter != null) {
                            if (getActivity() instanceof EntourageActivity) {
                                ((EntourageActivity) getActivity()).showProgressDialog(0);
                            }
                            presenter.deleteAccount();
                        }
                    }
                })
                .setNegativeButton(R.string.no, null);
        builder.show();
    }

    @OnClick(R.id.title_action_button)
    protected void onSaveButtonClicked() {
        EntourageEvents.logEvent(EntourageEvents.EVENT_USER_SAVE);
        // If we have an user fragment in the stack, let it handle the update
        UserFragment fragment = (UserFragment)getFragmentManager().findFragmentByTag(UserFragment.TAG);
        if (fragment != null) {
            fragment.saveAccount(editedUser);
        } else if (presenter != null) {
            //else we handle it
            presenter.updateUser(editedUser);
        }
    }

    @OnClick(R.id.user_photo_button)
    protected void onPhotoClicked() {
        PhotoChooseSourceFragment fragment = new PhotoChooseSourceFragment();
        fragment.show(getFragmentManager(), PhotoChooseSourceFragment.TAG);
    }

    @OnClick(R.id.user_add_association_button)
    public void onAddAssociationClicked() {
        EntourageEvents.logEvent(EntourageEvents.EVENT_USER_TOBADGE);
        UserEditPartnerFragment userEditPartnerFragment = new UserEditPartnerFragment();
        userEditPartnerFragment.show(getFragmentManager(), UserEditPartnerFragment.TAG);
    }

    @OnClick(R.id.user_notifications_layout)
    protected void onShowNotificationsSettingsClicked() {
        try {
            EntourageEvents.logEvent(EntourageEvents.EVENT_USER_TONOTIFICATIONS);
            Intent intent = new Intent();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                intent.setAction(Settings.ACTION_APP_NOTIFICATION_SETTINGS);
                intent.putExtra(Settings.EXTRA_APP_PACKAGE, getActivity().getPackageName());
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                intent.setAction("android.settings.APP_NOTIFICATION_SETTINGS");
                intent.putExtra("app_package", getActivity().getPackageName());
                intent.putExtra("app_uid", getActivity().getApplicationInfo().uid);
            } else {
                intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.addCategory(Intent.CATEGORY_DEFAULT);
                intent.setData(Uri.parse("package:" + getActivity().getPackageName()));
            }
            startActivity(intent);
        } catch (ActivityNotFoundException ex) {
            Timber.e(ex, "Cannot open Notifications Settings page");
        } catch (Exception ignored) {

        }
    }

    // ----------------------------------
    // Protected methods
    // ----------------------------------

    protected void scrollToOriginalPosition() {
        scrollView.scrollTo(0, scrollViewY);
    }

    // ----------------------------------
    // Private methods
    // ----------------------------------

    private void showEditProfile(int editType) {
        UserEditProfileFragment fragment = UserEditProfileFragment.newInstance(editType);
        fragment.show(getFragmentManager(), UserEditProfileFragment.TAG);
    }

    private void configureNotifications() {
        boolean areNotificationsEnabled = NotificationManagerCompat.from(getContext()).areNotificationsEnabled();
        userNotificationsStatusImage.setImageResource(areNotificationsEnabled ? R.drawable.verified  : R.drawable.not_verified);
    }

    // ----------------------------------
    // Events Handling
    // ----------------------------------

    @Subscribe
    public void userInfoUpdated(Events.OnUserInfoUpdatedEvent event) {
        if (getActivity() == null || getActivity().isFinishing()) {
            return;
        }

        User me = EntourageApplication.me(getActivity());
        editedUser.setAvatarURL(me.getAvatarURL());
        editedUser.setPartner(me.getPartner());
        editedUser.setAddress(me.getAddress());

        configureView();
    }

    @Subscribe
    public void onPartnerViewRequested(Events.OnPartnerViewRequestedEvent event) {
        if (getActivity() == null || getActivity().isFinishing()) {
            return;
        }
        if (event == null) {
            return;
        }
        if (event.getPartner() != null) {
            PartnerFragment.Companion.newInstance(event.getPartner()).show(getParentFragmentManager(), PartnerFragment.TAG);
        } else {
            PartnerFragment.Companion.newInstance(event.getPartnerId()).show(getParentFragmentManager(), PartnerFragment.TAG);
        }
    }

    // ----------------------------------
    // Presenter callbacks
    // ----------------------------------

    protected void onUserUpdated(User user) {
        if (getActivity() == null || getActivity().isFinishing()) {
            return;
        }
        if (user == null) {
            displayToast(getString(R.string.user_text_update_ko));
        }
        else {
            displayToast(getString(R.string.user_text_update_ok));
            dismiss();
        }
    }

    protected void onSaveNewPassword(String newPassword) {
        if (getActivity() == null || getActivity().isFinishing()) {
            return;
        }
        if (newPassword == null) {
            displayToast(getString(R.string.user_text_update_ko));
        }
        else {
            if (editedUser != null) {
                editedUser.setSmsCode(newPassword);
            }
            displayToast(getString(R.string.user_text_update_ok));
        }
        progressBar.setVisibility(View.GONE);
    }

    protected void onDeletedAccount(boolean success) {
        boolean hasActivity = getActivity() != null && !getActivity().isFinishing() && getActivity() instanceof EntourageActivity;
        if (hasActivity) {
            ((EntourageActivity) getActivity()).dismissProgressDialog();
        }
        if (success) {
            //logout and go back to login screen
            if (getActivity() instanceof MainActivity) {
                ((MainActivity) getActivity()).selectItem(R.id.action_logout);
            }
        }
        else {
            if (hasActivity) {
                Toast.makeText(getActivity(), R.string.user_delete_account_failure, Toast.LENGTH_SHORT).show();
            }
        }
    }

    // ----------------------------------
    // Edit Action Zone
    // ----------------------------------

    @OnClick(R.id.user_action_zone_button)
    protected void onActionZoneEditClicked() {
        if (getFragmentManager() == null) return;
        UserEditActionZoneFragment userEditActionZoneFragment = UserEditActionZoneFragment.newInstance(editedUser.getAddress());
        userEditActionZoneFragment.setFragmentListener(this);
        userEditActionZoneFragment.show(getFragmentManager(), UserEditActionZoneFragment.TAG);
    }

    @Override
    public void onUserEditActionZoneFragmentDismiss() {

    }

    @Override
    public void onUserEditActionZoneFragmentAddressSaved() {
        storeActionZone(false);
    }

    @Override
    public void onUserEditActionZoneFragmentIgnore() {
        storeActionZone(true) ;
    }

    public void storeActionZone(final boolean ignoreActionZone) {
        AuthenticationController authenticationController = EntourageApplication.get().getEntourageComponent().getAuthenticationController();
        UserPreferences userPreferences = authenticationController.getUserPreferences();
        if (userPreferences != null) {
            userPreferences.setIgnoringActionZone(ignoreActionZone);
            authenticationController.saveUserPreferences();
        }
        UserEditActionZoneFragment userEditActionZoneFragment = (UserEditActionZoneFragment)getFragmentManager().findFragmentByTag(UserEditActionZoneFragment.TAG);
        if (userEditActionZoneFragment != null && !userEditActionZoneFragment.isStateSaved()) {
            userEditActionZoneFragment.dismiss();
        }
    }
}

File: app/src/main/java/social/entourage/android/user/edit/UserEditFragment.kt
package social.entourage.android.user.edit

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.core.app.NotificationManagerCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.squareup.otto.Subscribe
import kotlinx.android.synthetic.main.fragment_user_edit.*
import kotlinx.android.synthetic.main.layout_view_title.*
import social.entourage.android.EntourageApplication
import social.entourage.android.EntourageComponent
import social.entourage.android.MainActivity
import social.entourage.android.R
import social.entourage.android.api.model.BaseOrganization
import social.entourage.android.api.model.User
import social.entourage.android.api.tape.Events.OnPartnerViewRequestedEvent
import social.entourage.android.api.tape.Events.OnUserInfoUpdatedEvent
import social.entourage.android.base.BaseActivity
import social.entourage.android.base.BaseDialogFragment
import social.entourage.android.tools.EntBus
import social.entourage.android.tools.log.AnalyticsEvents
import social.entourage.android.user.UserOrganizationsAdapter
import social.entourage.android.user.edit.partner.UserEditPartnerFragment
import social.entourage.android.user.edit.photo.ChoosePhotoFragment
import social.entourage.android.user.edit.place.UserEditActionZoneFragment
import social.entourage.android.user.edit.place.UserEditActionZoneFragment.FragmentListener
import social.entourage.android.user.partner.PartnerFragment
import timber.log.Timber
import java.util.*
import javax.inject.Inject

open class UserEditFragment  : BaseDialogFragment(), FragmentListener {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    internal var presenter: UserEditPresenter = UserEditPresenter(this)

    private var scrollViewY = 0

    private var organizationsAdapter: UserOrganizationsAdapter? = null
    var isShowAction = false
    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
           isShowAction = it.getBoolean(ARG_IS_SHOW_ACTION,false)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        AnalyticsEvents.logEvent(AnalyticsEvents.EVENT_SCREEN_09_2)
        return inflater.inflate(R.layout.fragment_user_edit, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        title_close_button?.setOnClickListener { onCloseButtonClicked() }
        user_firstname_layout?.setOnClickListener { onEditFirstname() }
        user_lastname_layout?.setOnClickListener { onEditFirstname() }
        user_email_layout?.setOnClickListener { onEditEmail() }
        user_password_layout?.setOnClickListener { onEditPassword() }
        user_about_edit_button?.setOnClickListener { onEditAboutClicked() }
        user_delete_account_button?.setOnClickListener { onDeleteAccountClicked() }
        title_action_button?.setOnClickListener { presenter.updateUser()  }
        user_photo_button?.setOnClickListener { onPhotoClicked() }
        user_add_association_button?.setOnClickListener { onAddAssociationClicked() }
        user_notifications_layout?.setOnClickListener { onShowNotificationsSettingsClicked() }

        ui_iv_action_zone1_mod?.setOnClickListener {
            onActionZoneEditClicked(false)
        }
        ui_iv_action_zone2_mod?.setOnClickListener {
            onActionZoneEditClicked(true)
        }

        ui_iv_action_zone2_delete?.setOnClickListener {
            presenter.deleteSecondaryAddress()
        }

        ui_profile_add_zone?.setOnClickListener {
            if(presenter.editedUser?.address!=null) {
                onActionAddSecondaryZone()
            } else { // if no first Zone is set, we set it first
                onActionZoneEditClicked(false)
            }
        }

        ui_iv_action_type_mod?.setOnClickListener {
           onActionSelectType()
        }

        initUserData()

        if (isShowAction) {
            onActionSelectType()
        }
    }

    override fun onStart() {
        super.onStart()
        configureNotifications()
        EntBus.register(this)
    }

    override fun onStop() {
        super.onStop()
        EntBus.unregister(this)
    }

    fun initUserData() {
        presenter.editedUser?.let {user->
            user_photo?.let { userPhoto ->
                user.avatarURL?.let { avatarURL->
                    Glide.with(this)
                            .load(Uri.parse(avatarURL))
                            .placeholder(R.drawable.ic_user_photo)
                            .circleCrop()
                            .into(userPhoto)
                } ?: run {
                    Glide.with(this)
                            .load(R.drawable.ic_user_photo)
                            .circleCrop()
                            .into(userPhoto)
                }
            }
            user_edit_firstname?.text = user.firstName
            user_edit_lastname?.text = user.lastName
            user_email?.text = user.email
            user_phone?.text = user.phone
            user_about?.text = user.about

            user_about_edit_button?.text = if (!user.about.isNullOrBlank()) getString(R.string.user_about_mod_button) else  getString(R.string.user_about_edit_button)

            val organizationList: MutableList<BaseOrganization> = ArrayList()
            user.partner?.let { organizationList.add(it) }
            user.organization?.let { organizationList.add(it) }
            if (organizationList.size > 0) {
                organizationsAdapter?.setOrganizationList(organizationList)
                        ?: run {
                    user_associations_view?.layoutManager = LinearLayoutManager(context)
                    organizationsAdapter = UserOrganizationsAdapter(organizationList)
                    user_associations_view?.adapter = organizationsAdapter
                }
                user_associations_layout?.visibility = View.VISIBLE
            }
            user.address?.let { address->
                ui_tv_action_zone1_title?.text = address.displayAddress
            } ?: run {
                ui_tv_action_zone1_title?.text = getString(R.string.user_edit_action_zone_button_no_address)
            }

            user.addressSecondary?.let {addressSecondary ->
                ui_tv_action_zone2_title?.text = addressSecondary.displayAddress
                ui_layout_2nd_zone?.visibility = View.VISIBLE
                ui_layout_add_zone?.visibility = View.GONE
            } ?: run {
                ui_layout_2nd_zone?.visibility = View.GONE
                ui_layout_add_zone?.visibility = View.VISIBLE
            }

            //Type
            ui_tv_action_type_desc?.text = when(user.goal) {
                User.USER_GOAL_NEIGHBOUR -> getString(R.string.onboard_type_choice_neighbour)
                User.USER_GOAL_ALONE -> getString(R.string.onboard_type_choice_alone)
                User.USER_GOAL_ASSO -> getString(R.string.onboard_type_choice_asso)
                else -> {getString(R.string.profile_action_not_selected) }
            }

            val interests = user.getFormattedInterests(requireContext())
            ui_tv_action_type_title?.text = String.format(getString(R.string.profile_activities),interests)
        }
    }

    private fun displayToast(message: String?) {
        activity?.let { Toast.makeText(it, message, Toast.LENGTH_SHORT).show() }
    }

    fun saveNewPassword(newPassword: String?) {
        presenter.saveNewPassword(newPassword ?: "").also { user_edit_progressBar?.visibility = View.VISIBLE }
    }

    // ----------------------------------
    // Buttons Handling
    // ----------------------------------
    private fun onCloseButtonClicked() {
        dismiss()
    }

    private fun onEditFirstname() {
        showEditProfile(UserEditProfileFragment.EDIT_NAME)
    }

    private fun onEditEmail() {
        showEditProfile(UserEditProfileFragment.EDIT_EMAIL)
    }

    private fun onEditPassword() {
        UserEditPasswordFragment().show(parentFragmentManager, UserEditPasswordFragment.TAG)
    }

    private fun onEditAboutClicked() {
        scrollView?.let { scrollViewY = it.scrollY }
        UserEditAboutFragment().show(parentFragmentManager, UserEditAboutFragment.TAG)
    }

    private fun onDeleteAccountClicked() {
        activity?.let {
            val builder = AlertDialog.Builder(it)
            builder.setMessage(R.string.user_delete_account_dialog)
                    .setPositiveButton(R.string.yes) { _, _ ->
                        (it as BaseActivity?)?.showProgressDialog(0)
                        presenter.deleteAccount()
                    }
                    .setNegativeButton(R.string.no, null)
            builder.show()
        }
    }

    private fun onPhotoClicked() {
        ChoosePhotoFragment.newInstance().show(parentFragmentManager, ChoosePhotoFragment.TAG)
    }

    fun onAddAssociationClicked() {
        AnalyticsEvents.logEvent(AnalyticsEvents.EVENT_USER_TOBADGE)
        UserEditPartnerFragment().show(parentFragmentManager, UserEditPartnerFragment.TAG)
    }

    private fun onShowNotificationsSettingsClicked() {
        activity?.let {
            try {
                AnalyticsEvents.logEvent(AnalyticsEvents.EVENT_USER_TONOTIFICATIONS)
                val intent = Intent()
                when {
                    Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> {
                        intent.action = Settings.ACTION_APP_NOTIFICATION_SETTINGS
                        intent.putExtra(Settings.EXTRA_APP_PACKAGE, it.packageName)
                    }
                    else -> {
                        intent.action = "android.settings.APP_NOTIFICATION_SETTINGS"
                        intent.putExtra("app_package", it.packageName)
                        intent.putExtra("app_uid", it.applicationInfo.uid)
                    }
                }
                startActivity(intent)
            } catch (ex: ActivityNotFoundException) {
                Timber.e(ex, "Cannot open Notifications Settings page")
            } catch (ignored: Exception) {
            }
        }
    }

    private fun onActionSelectType() {
         UserEditProfileType().show(parentFragmentManager, UserEditProfileType.TAG)
    }

    private fun onActionAddSecondaryZone() {
        val frag = UserEditActionZoneFragment.newInstance(null, true)
        frag.setupListener(this)
        frag.show(parentFragmentManager, UserEditActionZoneFragment.TAG)
    }

    // ----------------------------------
    // Protected methods
    // ----------------------------------
    fun scrollToOriginalPosition() {
        scrollView?.scrollTo(0, scrollViewY)
    }

    // ----------------------------------
    // Private methods
    // ----------------------------------
    private fun showEditProfile(editType: Int) {
        UserEditProfileFragment.newInstance(editType).show(parentFragmentManager, UserEditProfileFragment.TAG)
    }

    private fun configureNotifications() {
        val areNotificationsEnabled = NotificationManagerCompat.from(requireContext()).areNotificationsEnabled()
        user_notifications_image?.setImageResource(if (areNotificationsEnabled) R.drawable.verified else R.drawable.not_verified)
    }

    // ----------------------------------
    // Events Handling
    // ----------------------------------
    @Subscribe
    fun userInfoUpdated(event: OnUserInfoUpdatedEvent) {
        activity?.let {
            if (it.isFinishing) {
                return
            }
            presenter.initEditedUser()
        }
    }

    @Subscribe
    fun onPartnerViewRequested(event: OnPartnerViewRequestedEvent?) {
        if (activity?.isFinishing == true) {
            return
        }
        if (event == null) {
            return
        }
        PartnerFragment.newInstance(event.partner).show(parentFragmentManager, PartnerFragment.TAG)
    }

    // ----------------------------------
    // Presenter callbacks
    // ----------------------------------
    fun onUserUpdated(user: User) {
        if (activity?.isFinishing == true || this.isDetached) {
            return
        }
        try {
            displayToast(getString(R.string.user_text_update_ok))
            dismiss()
        } catch (e: IllegalStateException) {
            Timber.w(e)
        }
    }

    fun onUserUpdateError() {
        if (activity?.isFinishing == true || isDetached) {
            return
        }
        displayToast(getString(R.string.user_text_update_ko))
    }

    fun onSavePasswordError() {
        activity?.let {
            if (it.isFinishing) {
                return
            }
            displayToast(getString(R.string.user_text_update_ko))
            user_edit_progressBar?.visibility = View.GONE
        }
    }

    fun onSaveNewPassword() {
        activity?.let {
            if (it.isFinishing) {
                return
            }
            displayToast(getString(R.string.user_text_update_ok))
            user_edit_progressBar?.visibility = View.GONE
        }
    }

    fun onDeletedAccount(success: Boolean) {
        (activity as? BaseActivity)?.let {
            val hasActivity = !it.isFinishing
            if (hasActivity) {
                it.dismissProgressDialog()
            }
            if (success) {
                //logout and go back to login screen
                if (it is MainActivity) {
                    it.selectMenuProfileItem("logout")
                }
            } else if (hasActivity) {
                Toast.makeText(it, R.string.user_delete_account_failure, Toast.LENGTH_SHORT).show()
            }
        }
    }

    // ----------------------------------
    // Edit Action Zone
    // ----------------------------------
    private fun onActionZoneEditClicked(isSecondary: Boolean) {
        //ATTENTION both addresses can be null when they are not set yet !
        val frag = UserEditActionZoneFragment.newInstance(
                if (isSecondary) presenter.editedUser?.addressSecondary else presenter.editedUser?.address,
                isSecondary
        )
        frag.setupListener(this)
        frag.show(parentFragmentManager, UserEditActionZoneFragment.TAG)
    }

    override fun onUserEditActionZoneFragmentDismiss() {}
    override fun onUserEditActionZoneFragmentAddressSaved() {
        presenter.storeActionZone(false)
    }

    override fun onUserEditActionZoneFragmentIgnore() {
        presenter.storeActionZone(true)
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        const val TAG = "user_edit_fragment"
        const val ARG_IS_SHOW_ACTION = "isShowAction"
        fun newInstance(isShowAction:Boolean) =
                UserEditFragment().apply {
                    arguments = Bundle().apply {
                        putBoolean(ARG_IS_SHOW_ACTION,isShowAction)
                    }
                }
    }
}